code/renderers/vue3/src/render.ts

```diff
@@ -22,65 +30,118 @@ export const setup = (fn: (app: any) => void) => {
+  // create vue app for the story
+  const vueApp = createApp({
+    setup() {
+      storyContext.args = reactiveArgs;
+      reactiveState = reactive({ globals: storyContext.globals });
+      let rootElement: StoryFnVueReturnType = storyFn();
+
+      watch(
+        () => reactiveState.globals,
+        () => {
+          storyContext.globals = reactiveState.globals;
+          rootElement = storyFn();
+        }
+      );
+
+      return () => {
+        return h(rootElement, reactiveArgs);
+      };
+    },
+    mounted() {
+      map.set(canvasElement, {
+        vueApp,
+        reactiveArgs,
+      });
     },
   });
-
-  storybookApp.config.errorHandler = (e: unknown) => showException(e as Error);
-  setupFunction(storybookApp);
-  storybookApp.mount(canvasElement);
+  vueApp.config.errorHandler = (e: unknown) => showException(e as Error);
+  setupFunction(vueApp);
+  vueApp.mount(canvasElement);
 
   showMain();
   return () => {
-    teardown(storybookApp, canvasElement);
+    teardown(vueApp, canvasElement);
   };
 }
 
 /**
- * get the slots as functions to be rendered
- * @param props
+ * generate slots for default story without render function template
  * @param context
  */
 
-function getSlots(props: Args, context: StoryContext<VueRenderer, Args>) {
+function generateSlots(context: StoryContext<VueRenderer, Args>) {
   const { argTypes } = context;
-  const slots = Object.entries(props)
+  const slots = Object.entries(argTypes)
     .filter(([key, value]) => argTypes[key]?.table?.category === 'slots')
-    .map(([key, value]) => [key, typeof value === 'function' ? value : () => value]);
+    .map(([key, value]) => [
+      key,
+      () => {
+        const slotValue = context.args[key];
+        return typeof slotValue === 'function' ? h(slotValue) : slotValue;
+      },
+    ]);
+
+  return reactive(Object.fromEntries(slots));
+}
+
+/**
+ *  update the context args in case of decorators that change args
+ * @param storyFn
+ * @param storyContext
+ */
 
-  return Object.fromEntries(slots);
+function updateContextDecorator(
+  storyFn: PartialStoryFn<VueRenderer>,
+  storyContext: StoryContext<VueRenderer>
+) {
+  const storyDecorators: Set<DecoratorFunction<VueRenderer>> = (
+    storyContext.hooks as HooksContext<VueRenderer>
+  ).mountedDecorators;
+
+  if (storyDecorators && storyDecorators.size > 0) {
+    storyDecorators.forEach((decorator: DecoratorFunction<VueRenderer>) => {
+      try {
+        if (typeof decorator === 'function') {
+          decorator((u) => {
+            if (u && u.args && !u.globals) return storyFn();
+            return () => {};
+          }, storyContext);
+        }
+      } catch (e) {
+        // in case the decorator throws an error, we need to re-render the story
+        // mostly because of react hooks that are not allowed to be called conditionally
+        reactiveState.globals = storyContext.globals; // { ...storyContext.globals };
+      }
+    });
+  }
 }
```

Comment:
Can you explain what this is doing a little more? I'm a little puzzled about what's happening inside here.

Reply:
yes i have  vue is fine grained reactive, so prepareStory runs 1 time to create the Ui, however we may have decorators where global or in Story level some alters the args,  like 
```ts
 (storyFn: PartialStoryFn, context: StoryContext) => {
      const { argNames } = context.parameters;
      const object = argNames ? pick(context.args, argNames) : context.args;
      return storyFn({ args: { object } });
    },
  ],
```

so i need to run these to update my reactive context.
in case of Decorators that have react hooks will through an error  that i catch to update the context.glabals , if there is the change i will rerender the story so the react hooks got executed

Reply:
So if I am understanding correctly, the issue is that if a decorator:

 - Reads one arg
 - Sets or overrides another arg

then without this code the decorator will not re-run when the first arg changes, so the second arg will have the incorrect value? And the reason for that is that the way the `renderToDom`/`applyDecorators` function is setup, means that when the first arg changes, the decorator would not otherwise re-run?

I think this isn't the right solution to that problem. Here are some other issues I can see with the above off the top of my head:

1. Storybook hooks won't work (you said "react hooks" but I assume you mean SB hooks). What if the decorator relies on them?

3. What if the decorator changes the actual rendered output? Will that get dropped? What if the decorator changes some other field?

I also am not sure about the `if (u && u.args && !u.globals)` guard. What's the idea there?

I think this is a tricky problem, if I am understanding correctly (apologies if not). My feeling is the whole stack of decorators should re-run whenever args change. If that's fundamentally incompatible with Vue (or other renderers) we should probably rethink the contract of decorators or even make it different in these frameworks than React (i.e. not let them even read args). This approach seems fraught with problems.

Reply:
Hi T

> So if I am understanding correctly, the issue is that if a decorator:
> 
> * Reads one arg
> * Sets or overrides another arg
> 
> then without this code the decorator will not re-run when the first arg changes, so the second arg will have the incorrect value? And the reason for that is that the way the `renderToDom`/`applyDecorators` function is setup, means that when the first arg changes, the decorator would not otherwise re-run?
> 
> I think this isn't the right solution to that problem. Here are some other issues I can see with the above off the top of my head:
> 
> 1. Storybook hooks won't work (you said "react hooks" but I assume you mean SB hooks). What if the decorator relies on them?
> 2. What if the decorator changes the actual rendered output? Will that get dropped? What if the decorator changes some other field on the context?
> 
> I also am not sure about the `if (u && u.args && !u.globals)` guard. What's the idea there?
> 
> I think this is a tricky problem, if I am understanding correctly (apologies if not). My feeling is the whole stack of decorators should "properly" re-run whenever args change (treating the story's render function as the final decorator). If that's fundamentally incompatible with Vue (or other renderers) we should probably rethink the contract of decorators or even make it different in these frameworks than React (i.e. not let them even read args, or make a `useArgs` style API that forces Vue to re-render them). This approach seems fraught with problems.

Hi @tmeasday  i thought about your question "updated context should be passed to story?" and i managed to get it done, it helped removed a code i did not like cause it was just workout for me you can check my new code

Reply:
> > I think this is a tricky problem, if I am understanding correctly (apologies if not). My feeling is the whole stack of decorators should "properly" re-run whenever args change (treating the story's render function as the final decorator). If that's fundamentally incompatible with Vue (or other renderers) we should probably rethink the contract of decorators or even make it different in these frameworks than React (i.e. not let them even read args, or make a `useArgs` style API that forces Vue to re-render them). This approach seems fraught with problems.

Yep i agree that is why i needed your opinion. i could rely drop that workaround i have it done like you said just passing the the updated context,  this fix the vue decorators, and for decorators with hook, i'm watching context.globals is there is a change i trigger the vue app render  after recreating the VDom. ( react hooks works)